packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol

```diff
@@ -0,0 +1,201 @@
+    modifier notEntered() {
+        if (TransientContext.callDepth() == 0) revert NotEntered();
+        _;
+    }
+
+    /// @notice Retrieves the sender of the current cross domain message. If not entered, reverts.
+    /// @return Address of the sender of the current cross domain message.
+    function crossDomainMessageSender() external view notEntered returns (address) {
+        return address(uint160(TransientContext.get(CROSS_DOMAIN_MESSAGE_SENDER_SLOT)));
+    }
+
+    /// @notice Retrieves the source of the current cross domain message. If not entered, reverts.
+    /// @return Chain ID of the source of the current cross domain message.
+    function crossDomainMessageSource() external view notEntered returns (uint256) {
+        return TransientContext.get(CROSS_DOMAIN_MESSAGE_SOURCE_SLOT);
+    }
+
+    /// @notice Sends a message to some target address on a destination chain. Note that if the call always reverts,
+    ///         then the message will be unrelayable and any ETH sent will be permanently locked. The same will occur
+    ///         if the target on the other chain is considered unsafe (see the _isUnsafeTarget() function).
+    /// @param _destination Chain ID of the destination chain.
+    /// @param _target      Target contract or wallet address.
+    /// @param _message     Message payload to call target with.
+    function sendMessage(uint256 _destination, address _target, bytes calldata _message) external payable {
+        if (_destination == block.chainid) revert MessageDestinationSameChain();
+        if (_target == Predeploys.CROSS_L2_INBOX) revert MessageTargetCrossL2Inbox();
+        if (_target == Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) revert MessageTargetL2ToL2CrossDomainMessenger();
+
+        bytes memory data = abi.encodeCall(
+            L2ToL2CrossDomainMessenger.relayMessage,
+            (_destination, block.chainid, messageNonce(), msg.sender, _target, _message)
+        );
+        emit SentMessage(data);
+        msgNonce++;
+    }
+
+    /// @notice Relays a message that was sent by the other CrossDomainMessenger contract. Can only be executed via
+    ///         cross-chain call from the other messenger OR if the message was already received once and is currently
+    ///         being replayed.
+    /// @param _destination Chain ID of the destination chain.
+    /// @param _source      Chain ID of the source chain.
+    /// @param _nonce       Nonce of the message being relayed.
+    /// @param _sender      Address of the user who sent the message.
+    /// @param _target      Address that the message is targeted at.
+    /// @param _message     Message payload to call target with.
+    function relayMessage(
+        uint256 _destination,
+        uint256 _source,
+        uint256 _nonce,
+        address _sender,
+        address _target,
+        bytes memory _message
+    )
+        external
+        payable
+        reentrantAware
```

Comment:
I am going back and forth on this being able to accept reentrancy. Right now it follows the checks effects interactions pattern so i believe it should be safe with reentrancy. Its possible that disallowing explicit reentrancy and using a simple reentrancy lock could be simpler to reason about. The logic would be more like:

- set lock slot
- <do relayMessage>
- clear transient storage by relayMessage
- clear lock slot

This would still allow for multicall style patterns or 4337 bundlers to call this. I think the checks on the target, msg.sender and origin make it reentrant safe but still feels difficult to reason about fully

Reply:
Ok I think that we should not allow for reentrancy here as it gives a much more simple to reason about mental model for applications built on top of the messenger. Without it, every application on top would need to ensure that they are safe from reentrancy. For example the superchain erc20 would have to handle reentrancy if the messenger didn't provide it. There isn't an obvious usecase for allowing reentrancy and it seems like it will only result in bugs on top, so we should disallow it for simplicity. This prevents every application on top from needing to implement it themselves. It is possible that another messenger could exist that allows for reentrancy and if it becomes super useful, then we can reconsider this decision by adding an additional interface to this contract that allows for reentrency.